### PR TITLE
Split settings panels logic and add support for custom panels

### PR DESCRIPTION
**Pull request by dragunoff**
This PR:
- Splits `SettingsLogic` into chrome logic classes for each panel
- Splits `settings.yaml` into a wrapper for the settings window and containers for each panel
- Adds the ability to add custom settings panels

Why:
- `SettingsLogic` had become very large and difficult to maintain
- Mods could not add their own settings tabs

### CODE REVIEW

**Comment by pchote** [comment]
Good idea. How about putting these in a `Settings` subdir so they can stay grouped, like we have done for `Installation`, `Lobby`, etc?

**Comment by dragunoff** [comment]
Update: moved classes to a `Settings` subdir.

**Comment by dragunoff** [comment]
Update: rebased after #18973 and #18959

**Comment by pchote** [comment]
Can you please start bugging me on Discord every couple of days until I review this? We should try and merge this before anyone else wants to touch the settings menu.

File: OpenRA.Mods.Common/Widgets/Logic/Settings/SettingsLogic.cs
Comment (dragunoff) [inline OpenRA.Mods.Common/Widgets/Logic/Settings/SettingsLogic.cs]: I'm not sure if passing the whole instance here is a good idea. Only access to the method `RegisterSettingsPanel` is necessary but I wasn't sure how to do it.

File: OpenRA.Mods.Common/Widgets/Logic/Settings/SettingsLogic.cs
Comment (dragunoff) [inline OpenRA.Mods.Common/Widgets/Logic/Settings/SettingsLogic.cs]: This method keeps the same logic as before only now each separate chrome logic class registers itself.

File: mods/cnc/chrome/settings.yaml
Comment (dragunoff) [inline mods/cnc/chrome/settings.yaml:26]: Tab buttons are created from a template.

File: mods/cnc/chrome/settings.yaml
Comment (dragunoff) [inline mods/cnc/chrome/settings.yaml:33]: Tab panes are created from a template.

File: OpenRA.Mods.Common/Widgets/Logic/Settings/SettingsUtils.cs
Comment (dragunoff) [inline OpenRA.Mods.Common/Widgets/Logic/Settings/SettingsUtils.cs:17]: I moved the common methods for binding inputs to a static class.

**Review by dragunoff — COMMENTED** [review]
@pchote I added some comments over the code to guide the review. The diff is quite large but that's mostly because of splitting and moving things around.

File: OpenRA.Mods.Common/Widgets/Logic/Settings/SettingsLogic.cs
Comment (pchote) [reply OpenRA.Mods.Common/Widgets/Logic/Settings/SettingsLogic.cs]: We could pass `{ "registerPanel", (Action<string, string, Func<Widget, Action>, Func<Widget, Action>>)RegisterSettingsPanel },` instead and invoke it directly in the panels?

Comment (dragunoff) [reply OpenRA.Mods.Common/Widgets/Logic/Settings/SettingsLogic.cs]: Yes, that works.

**Comment by dragunoff** [comment]
Update: Cleaned up unnecessary usings and constructor arguments and addressed https://github.com/OpenRA/OpenRA/pull/18948#discussion_r556800781

**Review by pchote — COMMENTED** [review]
Looks good overall. I diffed the split files versus the original and have confirmed that nothing has been unexpectedly dropped, or changed during rebases etc.

IMO having the panels reach into SettingsLogic to modify NeedsRestart isn't great - what do you think about doing this instead: ccef39683a5327f033a1f0129a06c30a2dc88973 ?

File: mods/cnc/chrome/settings.yaml
Comment (pchote) [inline mods/cnc/chrome/settings.yaml]: These keys don't seem to be used at all?

How about using a straight dictionary?
```yaml
Panels:
	DISPLAY_PANEL: Display
	AUDIO_PANEL: Audio
	INPUT_PANEL: Input
	HOTKEYS_PANEL: Hotkeys
	ADVANCED_PANEL: Advanced
```
?

File: OpenRA.Mods.Common/Widgets/Logic/Settings/SettingsLogic.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Settings/SettingsLogic.cs]: Cleaner if this defaulted to null?

File: mods/cnc/chrome/settings.yaml
Comment (dragunoff) [reply mods/cnc/chrome/settings.yaml]: Yes, this is more concise and straight-forward :+1:

**Comment by dragunoff** [comment]
Update:
- rebased
- cherry-picked ccef396
- applied other fixups (see commit messages)

I will squash the commits once this is approved.

**Review by pchote — DISMISSED** [review]
LGTM once the fixups are squashed.

**Comment by dragunoff** [comment]
Update: squashed fixups.

**Comment by pchote** [comment]
pinging for a second review.

**Review by abcdefg30 — APPROVED** [review]
Rebased and updated the dates to 2021.

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d8020a6fb402750b28416c36fa6b6b88158c4bc6...a7116c1232682fea806f31508398f5360d16130a)

**Pull request merged**
